fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -1416,16 +1442,49 @@ struct RocksDBKeyValueStore : IKeyValueStore {
 		}
 	}
 
+	std::vector<MutationRef> getDisposeShardMutation(KeyRangeRef range) override {
+		std::vector<MutationRef> refs;
+		auto shard = shardMap.rangeContaining(range.begin);
+		if (shard.range() != range) {
+			TraceEvent(SevError, "RocksDB")
+			    .detail("Method", "DisposeShard")
+			    .detail("Error", "ShardMismatch")
+			    .detail("Range", range.begin.toString() + " : " + range.end.toString())
+			    .detail("ExistingRange", shard.range().begin.toString() + " : " + shard.range().end.toString());
+			return refs;
+		}
+
+		if (!shard.value()) {
+			TraceEvent(SevError, "RocksDB").detail("Method", "DisposeShard").detail("Error", "Shard not exist.");
+			return refs;
+		}
+
+		// This may cause multiple fragments in metadata, but will not affect correctness.
+		refs.push_back(MutationRef(MutationRef::ClearRange,
+		                           persistShardMappingPrefix.toString() + range.begin.toString(),
+		                           persistShardMappingPrefix.toString() + range.end.toString()));
+		refs.push_back(MutationRef(MutationRef::SetValue,
+		                           persistShardMappingPrefix.toString() + range.begin.toString(),
+		                           LiteralStringRef("")));
+
+		// TODO: should we include sys key range here?
+		if (range.end < defaultShardRange.begin) {
+			auto it = shardMap.rangeContaining(range.end);
+			refs.push_back(MutationRef(MutationRef::SetValue,
+			                           persistShardMappingPrefix.toString() + range.end.toString(),
+			                           it.value() == nullptr ? "" : it.value()->GetName()));
+		}
+		deletePendingShards->push_back(shard.value());
+
+		// The shard will stay in shardMap.
+		return std::move(refs);
+	}
+
 	Future<Void> disposeShard(KeyRangeRef range) override {
 		if (!SERVER_KNOBS->ROCKSDB_ENABLE_SHARDING) {
 			return Void();
 		}
 
-		if (writeBatches == nullptr) {
-			writeBatches.reset(new std::unordered_map<DB, rocksdb::WriteBatch>());
-			systemMutationBatch.reset(new rocksdb::WriteBatch());
-		}
-
 		auto shard = shardMap.rangeContaining(range.begin);
 		if (shard.range() != range) {
```

Comment:
Sometimes, we only want to drop a sub-range of an existing shard, or drop multiple shard if range covers more than one, so shard.range() == range may not hold.

Also, in case of a sub-shard, no rocks will be added to deletePending, and multiple rocks if it covers more than one shard.